### ISSUE TITLE
Document returned from promise won't save

### ISSUE THREAD

**Issue by Jks15063** [Automattic/mongoose#2948]
Issue only crops up after upgrade to version 4.  I can consistently downgrade to 3.8 and watch the save start working, and upgrade to 4 and see it stop working.

````
function findADoc() {
    return newTest.findOne({}).exec();
}

findADoc()
.then(function(doc) {
    console.log(err, doc);
    doc.set('foo', 'wut');
    doc.save(function(error) {
        console.log('->', error);
    });
});
```

Turned on logging, `.then` is hit, doc is logged to console, but save is never fired.
Something to do with being passed through a promise chain, documents returned to a callback will save.
For example:
```
newTest.findOne({}, function(err, doc) {
    doc.set('foo', 'wut');
    doc.save(function(error) {
        console.log('->', error);
    });
});
```
The above still works as expected.

Tried various combinations of wrapping it with Q:
`return Q(newTest.findOne({}).exec());`
`Q.when(findADoc)...`
I always get the doc, but save does not fire.

**Comment by vkarpov15** [comment]
The following script works for me in 4.0.3

```javascript
var mongoose = require('mongoose');
mongoose.set('debug', true);
var util = require('util');
var assert = require('assert');

mongoose.connect('mongodb://localhost:27017/gh2948');

var UserSchema = new mongoose.Schema({
  name: { type: String }
});

var User = mongoose.model('gh2948', UserSchema);

function findADoc() {
    return User.findOne({}).exec();
}

User.create({ name: 'val' }, function(error) {
  assert.ifError(error);
  findADoc().
    then(function(doc) {
      console.log('Got doc: ' + require('util').inspect(doc));
      doc.set('name', 'test');
      doc.save(function(error) {
        assert.ifError(error);
        process.exit(0);
      });
  });
});
```

Can you verify this works for you, and, if so, provide me a standalone script that reproduces the issue you're experiencing?

**Comment by vkarpov15** [comment]
No activity for a while, please re-open if this is still an issue.

**Comment by Jks15063** [comment]
Ah yea forgot to get back to you on this.  So, I'm still not completely sure what was causing this.  I narrowed it down to a a few lines of code that was a for loop pushing and popping values into a couple of arrays on a document before saving it.  Without that loop the save worked fine on 4.x, and in mongoose 3.8 the loop didn't break the save.  I refactored that bit to be a little cleaner and it now works with 4.x.  

I guess it could be some edge case I didn't accurately recreate when I tried again to recreate the behavior.  I currently cannot recreate so I'm not going to re open this.
